posthog/temporal/data_modeling/workflows/execute_dag.py

```diff
@@ -0,0 +1,337 @@
+                scheduled_nodes=0,
+                successful_nodes=0,
+                failed_nodes=0,
+                skipped_nodes=0,
+                duration_seconds=(end_time - start_time).total_seconds(),
+                node_results=[],
+            )
+
+        edge_lookup = _get_edge_lookup(dag_structure.edges)
+        levels = _dag_execution_levels(inputs.team_id, inputs.dag_id, executable_nodes, edge_lookup)
+
+        temporalio.workflow.logger.info(
+            "DAG execution levels",
+            extra={
+                "levels": levels,
+                "num_levels": len(levels),
+                "nodes_per_level": [len(level) for level in levels],
+                **inputs.properties_to_log,
+            },
+        )
+
+        node_results: list[NodeResult] = []
+        failed_node_set: set[str] = set()
+        downstreams = _get_downstream_lookup(edge_lookup)
+        for i, level in enumerate(levels):
+            temporalio.workflow.logger.info(
+                f"Executing level {i + 1}/{len(levels)}",
+                extra={"nodes": level, **inputs.properties_to_log},
+            )
+            execute_nodes = []
+            skip_nodes = []
+            for node_id in level:
+                should_skip = False
+                skip_reason = None
+                for failed_id in failed_node_set:
+                    if node_id in downstreams[failed_id]:
+                        should_skip = True
+                        skip_reason = f"Upstream node {failed_id} failed"
+                        break
+                if should_skip:
+                    skip_nodes.append((node_id, skip_reason))
+                else:
+                    execute_nodes.append(node_id)
+
+            for node_id, skip_reason in skip_nodes:
+                node_results.append(
+                    NodeResult(
+                        node_id=node_id,
+                        success=False,
+                        skipped=True,
+                        skip_reason=skip_reason,
+                    )
+                )
+
+            if not execute_nodes:
+                continue
+
+            # execute child workflows in parallel for this level
+            child_handles = []
+            for node_id in execute_nodes:
+                handle = await temporalio.workflow.start_child_workflow(
+                    MaterializeViewWorkflow.run,
+                    MaterializeViewWorkflowInputs(
+                        team_id=inputs.team_id,
+                        dag_id=inputs.dag_id,
+                        node_id=node_id,
+                    ),
+                    id=f"materialize-{inputs.dag_id}-{node_id}-{temporalio.workflow.now().isoformat()}",
+                    retry_policy=temporalio.common.RetryPolicy(
+                        maximum_attempts=1,  # retries handled within child workflow
+                    ),
+                )
+                child_handles.append((node_id, handle))
```

Comment:
I'm semi-surprised to see us using child _workflows_ for this instead of activities - we lose a level of retryability and observability here to some extent, I think (when compared to running an activity). There's a _bunch_ of logic going on in this workflow, and so im worried about how we recover from failures/errors (both transient and not)